File: lib/urielm_web/plugs/theme.ex
defmodule UrielmWeb.Plugs.Theme do
  @moduledoc """
  Reads the preferred theme from cookie and assigns it for layouts.

  Normalizes persisted theme preferences to the supported Tokyo theme pair.
  """
  import Plug.Conn

  @behaviour Plug

  @cookie "phx_theme"
  @light_themes ~w(tokyo-day light catppuccin-latte github-light cupcake bumblebee emerald corporate retro valentine garden lofi pastel fantasy wireframe cmyk autumn acid lemonade winter)

  @impl true
  def init(opts), do: opts

  @impl true
  def call(conn, _opts) do
    theme = normalize_theme(conn.cookies[@cookie])
    assign(conn, :theme, theme)
  end

  defp normalize_theme(theme) when theme in @light_themes, do: "tokyo-day"
  defp normalize_theme(_theme), do: "tokyo-night"
end

File: assets/js/theme.js
export const DARK_THEME = "tokyo-night"
export const LIGHT_THEME = "tokyo-day"

const LIGHT_LEGACY_THEMES = new Set([
  LIGHT_THEME,
  "light",
  "catppuccin-latte",
  "github-light",
  "cupcake",
  "bumblebee",
  "emerald",
  "corporate",
  "retro",
  "valentine",
  "garden",
  "lofi",
  "pastel",
  "fantasy",
  "wireframe",
  "cmyk",
  "autumn",
  "acid",
  "lemonade",
  "winter"
])

export function normalizeTheme(theme) {
  return LIGHT_LEGACY_THEMES.has(theme) ? LIGHT_THEME : DARK_THEME
}

export function oppositeTheme(theme) {
  return normalizeTheme(theme) === DARK_THEME ? LIGHT_THEME : DARK_THEME
}
